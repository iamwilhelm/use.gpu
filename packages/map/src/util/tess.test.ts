import { cutPolygon, cutRing, assembleCutRing, getRingArea, pointInRing } from './tess';

describe("tesselate", () => {

  it("cuts ring left", () => {

    const ring = [
      [  0,   0],
      [100,   0],
      [100, 100],
      [  0, 100],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);
    expect(cut).toMatchSnapshot();
  });

  it("cuts ring right", () => {

    const ring = [
      [  0,   0],
      [100,   0],
      [100, 100],
      [  0, 100],
    ];

    const nx = -1;
    const ny = 0;
    const d = -50;

    const cut = cutRing(ring, nx, ny, d);
    expect(cut).toMatchSnapshot();

  });

  it("cuts ring on vertex left", () => {

    const ring = [
      [  0,   0],
      [ 50,   0],
      [100,   0],
      [100, 100],
      [ 50, 100],
      [  0, 100],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);
    expect(cut).toMatchSnapshot();
  });

  it("cuts ring on vertex right", () => {

    const ring = [
      [  0,   0],
      [ 50,   0],
      [100,   0],
      [100, 100],
      [ 50, 100],
      [  0, 100],
    ];

    const nx = -1;
    const ny = 0;
    const d = -50;

    const cut = cutRing(ring, nx, ny, d);
    expect(cut).toMatchSnapshot();
  });

  it("cuts U", () => {

    const ring = [
      [  0,   0],
      [100,   0],
      [100,  25],
      [ 25,  25],
      [ 25,  75],
      [100,  75],
      [100, 100],
      [  0, 100],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);
    expect(cut).toMatchSnapshot();

  });

  it("cuts U left with multiple vertices on edge", () => {

    const ring = [
      [  0,   0],
      [ 50,   0],
      [ 50,  10],
      [ 50,  25],
      [ 25,  25],
      [ 25,  75],
      [100,  75],
      [100, 100],
      [  0, 100],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);
    expect(cut).toMatchSnapshot();

  });

  it("cuts U right with multiple vertices on edge", () => {

    const ring = [
      [  0,   0],
      [ 50,   0],
      [ 50,  10],
      [ 50,  25],
      [ 25,  25],
      [ 25,  75],
      [100,  75],
      [100, 100],
      [  0, 100],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);
    expect(cut).toMatchSnapshot();

  });

  it("assembles U pieces", () => {

    const ring = [
      [  0,   0],
      [100,   0],
      [100,  25],
      [ 25,  25],
      [ 25,  75],
      [100,  75],
      [100, 100],
      [  0, 100],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);
    const polygon = assembleCutRing(cut, nx, ny);

    expect(polygon.map(getRingArea)).toMatchSnapshot();
    expect(polygon).toMatchSnapshot();

  });

  it("assembles U piece left", () => {

    const ring = [
      [  0,   0],
      [100,   0],
      [100,  25],
      [ 25,  25],
      [ 25,  75],
      [100,  75],
      [100, 100],
      [  0, 100],
    ];

    const nx = -1;
    const ny = 0;
    const d = -50;

    const cut = cutRing(ring, nx, ny, d);

    const polygon = assembleCutRing(cut, nx, ny);
    expect(polygon.map(getRingArea)).toMatchSnapshot();
    expect(polygon).toMatchSnapshot();

  });

  it("assembles U piece right", () => {

    const ring = [
      [100,   0],
      [  0,   0],
      [  0,  25],
      [ 75,  25],
      [ 75,  75],
      [  0,  75],
      [  0, 100],
      [100, 100],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);

    const polygon = assembleCutRing(cut, nx, ny);
    expect(polygon.map(getRingArea)).toMatchSnapshot();
    expect(polygon).toMatchSnapshot();

  });

  it("assembles spiral piece", () => {

    const ring = [
      [  0,   0],
      [100,   0],
      [100, 100],
      [  0, 100],
      [  0,  25],
      [ 75,  25],
      [ 75,  75],
      [ 25,  75],
      [ 25,  50],
      [ 60,  50],
      [ 60,  55],
      [ 30,  55],
      [ 30,  70],
      [ 70,  70],
      [ 70,  30],
      [  5,  30],
      [  5,  95],
      [ 95,  95],
      [ 95,   5],
      [  0,   5],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = cutRing(ring, nx, ny, d);

    const polygon = assembleCutRing(cut, nx, ny);
    expect(polygon.map(getRingArea)).toMatchSnapshot();
    expect(polygon).toMatchSnapshot();

  });

  it("assembles complex piece", () => {

    const rings = [
      [
        [  0,   0],
        [100,   0],
        [100, 100],
        [  0, 100],
        [  0,  25],
        [ 75,  25],
        [ 75,  75],
        [ 25,  75],
        [ 25,  50],
        [ 60,  50],
        [ 60,  55],
        [ 30,  55],
        [ 30,  70],
        [ 70,  70],
        [ 70,  30],
        [  5,  30],
        [  5,  95],
        [ 95,  95],
        [ 95,   5],
        [  0,   5],
      ],
      [
        [45, 51],
        [45, 52],
        [55, 52],
        [55, 51],
      ],
      [
        [45, 53],
        [45, 54],
        [55, 54],
        [55, 53],
      ],
      [
        [45, 1],
        [45, 4],
        [55, 4],
        [55, 1],
      ],
      [
        [54, 2],
        [54, 3],
        [46, 3],
        [46, 2],
      ],
    ];

    const nx = 1;
    const ny = 0;
    const d = 50;

    const cut = rings.flatMap(r => cutRing(r, nx, ny, d));
    const polygon = assembleCutRing(cut, nx, ny);

    expect(polygon.map(getRingArea)).toMatchSnapshot();
    expect(polygon).toMatchSnapshot();

  });

  it('cut polygon', () => {
    const rings = [
      [
        [10, 10],
        [50, 10],
        [40, 40],
        [10, 50],
      ],
      [
        [20, 20],
        [20, 25],
        [25, 25],
        [25, 20],
      ],
    ];

    const nx = 1;
    const ny = 0;
    const d = 22;

    const polygons = cutPolygon(rings, nx, ny, d);

    expect(polygons.map(polygon => polygon.map(getRingArea))).toMatchSnapshot();
    expect(polygons).toMatchSnapshot();

  });

  it('cut polygon beside hole', () => {
    const rings = [
      [
        [10, 10],
        [50, 10],
        [40, 40],
        [10, 50],
      ],
      [
        [20, 20],
        [20, 25],
        [25, 25],
        [25, 20],
      ],
      [
        [18, 20],
        [18, 25],
        [19, 25],
        [19, 20],
      ],
      [
        [16, 20],
        [16, 25],
        [17, 25],
        [17, 20],
      ],
    ];

    const nx = 1;
    const ny = 0;
    const d = 12;

    const polygons = cutPolygon(rings, nx, ny, d);

    expect(polygons.map(polygon => polygon.map(getRingArea))).toMatchSnapshot();
    expect(polygons).toMatchSnapshot();

  });

  it('point in ring', () => {
    const ring = [
      [1, 1],
      [4, 1],
      [3, 3],
      [1, 2],
    ];

    expect(pointInRing(ring, [1.5, 1.5])).toBe(true);
    expect(pointInRing(ring, [-1.5, 1.5])).toBe(false);
  });


});
