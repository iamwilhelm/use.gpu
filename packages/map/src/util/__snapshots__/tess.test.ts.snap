// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tesselate assembles U piece left 1`] = `
Array [
  3750,
]
`;

exports[`tesselate assembles U piece left 2`] = `
Array [
  Array [
    Array [
      50,
      25,
    ],
    Array [
      25,
      25,
    ],
    Array [
      25,
      75,
    ],
    Array [
      50,
      75,
    ],
    Array [
      50,
      100,
    ],
    Array [
      0,
      100,
    ],
    Array [
      0,
      0,
    ],
    Array [
      50,
      0,
    ],
  ],
]
`;

exports[`tesselate assembles U piece right 1`] = `
Array [
  -3750,
]
`;

exports[`tesselate assembles U piece right 2`] = `
Array [
  Array [
    Array [
      50,
      25,
    ],
    Array [
      75,
      25,
    ],
    Array [
      75,
      75,
    ],
    Array [
      50,
      75,
    ],
    Array [
      50,
      100,
    ],
    Array [
      100,
      100,
    ],
    Array [
      100,
      0,
    ],
    Array [
      50,
      0,
    ],
  ],
]
`;

exports[`tesselate assembles U pieces 1`] = `
Array [
  1250,
  1250,
]
`;

exports[`tesselate assembles U pieces 2`] = `
Array [
  Array [
    Array [
      50,
      0,
    ],
    Array [
      100,
      0,
    ],
    Array [
      100,
      25,
    ],
    Array [
      50,
      25,
    ],
  ],
  Array [
    Array [
      50,
      75,
    ],
    Array [
      100,
      75,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
]
`;

exports[`tesselate assembles complex piece 1`] = `
Array [
  935,
  4,
  450,
  40,
]
`;

exports[`tesselate assembles complex piece 2`] = `
Array [
  Array [
    Array [
      50,
      95,
    ],
    Array [
      95,
      95,
    ],
    Array [
      95,
      5,
    ],
    Array [
      50,
      5,
    ],
    Array [
      50,
      4,
    ],
    Array [
      55,
      4,
    ],
    Array [
      55,
      1,
    ],
    Array [
      50,
      1,
    ],
    Array [
      50,
      0,
    ],
    Array [
      100,
      0,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
  Array [
    Array [
      50,
      2,
    ],
    Array [
      54,
      2,
    ],
    Array [
      54,
      3,
    ],
    Array [
      50,
      3,
    ],
  ],
  Array [
    Array [
      50,
      70,
    ],
    Array [
      70,
      70,
    ],
    Array [
      70,
      30,
    ],
    Array [
      50,
      30,
    ],
    Array [
      50,
      25,
    ],
    Array [
      75,
      25,
    ],
    Array [
      75,
      75,
    ],
    Array [
      50,
      75,
    ],
  ],
  Array [
    Array [
      50,
      54,
    ],
    Array [
      55,
      54,
    ],
    Array [
      55,
      53,
    ],
    Array [
      50,
      53,
    ],
    Array [
      50,
      52,
    ],
    Array [
      55,
      52,
    ],
    Array [
      55,
      51,
    ],
    Array [
      50,
      51,
    ],
    Array [
      50,
      50,
    ],
    Array [
      60,
      50,
    ],
    Array [
      60,
      55,
    ],
    Array [
      50,
      55,
    ],
  ],
]
`;

exports[`tesselate assembles spiral piece 1`] = `
Array [
  950,
  450,
  50,
]
`;

exports[`tesselate assembles spiral piece 2`] = `
Array [
  Array [
    Array [
      50,
      95,
    ],
    Array [
      95,
      95,
    ],
    Array [
      95,
      5,
    ],
    Array [
      50,
      5,
    ],
    Array [
      50,
      0,
    ],
    Array [
      100,
      0,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
  Array [
    Array [
      50,
      70,
    ],
    Array [
      70,
      70,
    ],
    Array [
      70,
      30,
    ],
    Array [
      50,
      30,
    ],
    Array [
      50,
      25,
    ],
    Array [
      75,
      25,
    ],
    Array [
      75,
      75,
    ],
    Array [
      50,
      75,
    ],
  ],
  Array [
    Array [
      50,
      50,
    ],
    Array [
      60,
      50,
    ],
    Array [
      60,
      55,
    ],
    Array [
      50,
      55,
    ],
  ],
]
`;

exports[`tesselate cut polygon 1`] = `
Array [
  Array [
    729,
  ],
]
`;

exports[`tesselate cut polygon 2`] = `
Array [
  Array [
    Array [
      Array [
        22,
        25,
      ],
      Array [
        25,
        25,
      ],
      Array [
        25,
        20,
      ],
      Array [
        22,
        20,
      ],
      Array [
        22,
        10,
      ],
      Array [
        50,
        10,
      ],
      Array [
        40,
        40,
      ],
      Array [
        22,
        46,
      ],
    ],
  ],
]
`;

exports[`tesselate cuts U 1`] = `
Array [
  Array [
    Array [
      50,
      0,
    ],
    Array [
      100,
      0,
    ],
    Array [
      100,
      25,
    ],
    Array [
      50,
      25,
    ],
  ],
  Array [
    Array [
      50,
      75,
    ],
    Array [
      100,
      75,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
]
`;

exports[`tesselate cuts U left with multiple vertices on edge 1`] = `
Array [
  Array [
    Array [
      50,
      75,
    ],
    Array [
      100,
      75,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
]
`;

exports[`tesselate cuts U right with multiple vertices on edge 1`] = `
Array [
  Array [
    Array [
      50,
      75,
    ],
    Array [
      100,
      75,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
]
`;

exports[`tesselate cuts ring left 1`] = `
Array [
  Array [
    Array [
      50,
      0,
    ],
    Array [
      100,
      0,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
]
`;

exports[`tesselate cuts ring on vertex left 1`] = `
Array [
  Array [
    Array [
      50,
      0,
    ],
    Array [
      100,
      0,
    ],
    Array [
      100,
      100,
    ],
    Array [
      50,
      100,
    ],
  ],
]
`;

exports[`tesselate cuts ring on vertex right 1`] = `
Array [
  Array [
    Array [
      50,
      100,
    ],
    Array [
      0,
      100,
    ],
    Array [
      0,
      0,
    ],
    Array [
      50,
      0,
    ],
  ],
]
`;

exports[`tesselate cuts ring right 1`] = `
Array [
  Array [
    Array [
      50,
      100,
    ],
    Array [
      0,
      100,
    ],
    Array [
      0,
      0,
    ],
    Array [
      50,
      0,
    ],
  ],
]
`;
